### PR TITLE
support integer-range type in min/max

### PR DESCRIPTION
**Pull request by staabm**
rewrite `min($x, $y)` as `$x < $y ? $x : $y`
implemented the integer-range min/max support based on the idea of ondrey in https://github.com/phpstan/phpstan-src/pull/604#discussion_r689440018

closes https://github.com/phpstan/phpstan/issues/5398

this fixes the reported issue, but I can think of more complex scenarios which are not yet covered, e.g.

```php
	/**
	 * @param int<0, 10> $range1
	 * @param int<3, 9> $range2
	 * @param int<4, 5> $range3
	 * @param int $int
	 */
	public function complexRanges($range1, $range2, $range3, $int) {
		assertType('int', min($range1, $range2, $int));
		assertType('int', max($range1, $range2, $int));

		assertType('int<0, 5>', min($range1, $range2, $range3));
		assertType('int<4, 10>', max($range1, $range2, $range3));
	}
```

but I think these should be covered in a separate PR, as they are rare and this PR - as is - already covers a lot of real world code

closes https://github.com/phpstan/phpstan-src/pull/604

### CODE REVIEW

File: tests/PHPStan/Analyser/AnalyserIntegrationTest.php [hunk 1]
```diff
@@ -278,9 +278,10 @@ public function testBug3379(): void
 			$this->markTestSkipped('Test requires static reflection');
 		}
 		$errors = $this->runAnalyse(__DIR__ . '/data/bug-3379.php');
-		$this->assertCount(2, $errors);
+		$this->assertCount(3, $errors);
 		$this->assertSame('Constant SOME_UNKNOWN_CONST not found.', $errors[0]->getMessage());
 		$this->assertSame('Reflection error: Could not locate constant "SOME_UNKNOWN_CONST" while evaluating expression in Bug3379\Foo at line 8', $errors[1]->getMessage());
+		$this->assertSame('Reflection error: Could not locate constant "SOME_UNKNOWN_CONST" while evaluating expression in Bug3379\Foo at line 8', $errors[2]->getMessage());
```
Comment (staabm) [inline tests/PHPStan/Analyser/AnalyserIntegrationTest.php:284]: atm I have no clue, why this additional error is coming up

Comment (staabm) [reply tests/PHPStan/Analyser/AnalyserIntegrationTest.php:284]: ahh just realized this error is also present on the latest master branch:
https://github.com/phpstan/phpstan-src/runs/3374000058

therefore it should not be related to the changes from this PR.

Comment (staabm) [reply tests/PHPStan/Analyser/AnalyserIntegrationTest.php]: Removed the hack. This should be good to go

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php [hunk 2]
```diff
@@ -64,6 +66,31 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 			return new ErrorType();
 		}
 
+		// rewrite min($x, $y) as $x < $y ? $x : $y
+		// in case we don't handle arrays, which have different semantics
```
Comment (staabm) [inline src/Type/Php/MinMaxFunctionReturnTypeExtension.php:70]:
```suggestion
		// we don't handle arrays, which have different semantics
```

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php [hunk 3]
```diff
@@ -64,6 +66,31 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
...
+				} elseif ($functionName === 'max') {
+					return $scope->getType(new Ternary(
+						new Smaller($args[0]->value, $args[1]->value),
+						$args[1]->value,
+						$args[0]->value
+					));
+				}
+			}
+		}
+
```
Comment (ondrejmirtes) [inline src/Type/Php/MinMaxFunctionReturnTypeExtension.php:93]: Isn't it possible that we could delete some code that's already in the extension thanks to this logic?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/MinMaxFunctionReturnTypeExtension.php (hunk 3 above)
Comment (staabm) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php:93]: the code above only handles the 1-arg case and the code below handles the any-number-of-args case.

I guess we can drop it, after we push the 

```
return $scope->getType(new Ternary(
						new Smaller($args[0]->value, $args[1]->value),
						$args[1]->value,
						$args[0]->value
					));
```

approach into a direction where it supports any number of args (using `BooleanAnd` and `BooleanOr`) as you suggested in https://github.com/phpstan/phpstan-src/pull/604#discussion_r689521573

but IMO this should be a separate PR

Comment (ondrejmirtes) [reply src/Type/Php/MinMaxFunctionReturnTypeExtension.php:93]: Cool :)

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**